### ISSUE TITLE
UV upgrades packages inherited with --system-site-packages

### ISSUE THREAD

**Issue by lkoelman** [astral-sh/uv#4466]
Hi there, I'm encountering the following behaviour where uv deviates from `pip` and is causing one of our builds to fail: when you create a virtual environment with `--system-site-packages`, and then install a package without specifying a version, UV chooses the upgrade path (install the latest version in the virtual environment), whereas pip just uses the older version inherited from the base environment.

How to reproduce:

```bash
# (in your base python environment, I used conda)
pip install 'numpy==1.27.0'
pip install uv

python -m venv .venv
source .venv/bin/activate
# install without verison requirement
uv pip install numpy
# ==> this installs numpy==2.0.0

# install using pip, without version requirements
pip install numpy
# ==> this doesn't install anything, expected behaviour
```

**Comment by zanieb** [comment]
Hi. Is `--system-site-packages` missing from your reproduction here?

**Comment by charliermarsh** [comment]
(I think this is roughly a known problem, since we don't look at the full `sys.path`.)

**Comment by charliermarsh** [comment]
(Same as #2500.)

**Comment by lkoelman** [comment]
> Hi. Is `--system-site-packages` missing from your reproduction here?

yes it was missing, I updated it in my original post

**Comment by zanieb** [comment]
Yeah I think the root cause is the same as #2500 — maybe we need a separate issue that tracks that functionality. Otherwise it seems reasonable to keep this open to track this effect.

**Comment by charliermarsh** [comment]
It has an effect on `python` itself -- your interpreter will have access to the system site packages. But we don't respect it in `uv`.

**Comment by J3ronimo** [comment]
Ah, understood.

We're making great use of this "diff install" behavior of pip to minimize venv sizes in our test suite. Each test needs numpy, matplotlib, scipy etc, which are provided through a rich base install (WinPython, similarly Anaconda) and the dependency requirements of the tests are tried to make best use of what's already in there. This way we can typically reduce >200MB full installs to <10MB, which adds up with every test and version thereof.

Is there another feature in uv that would allow this kind of diff install scenario in order save disk space? Do you plan to add it at some point (or have you already decided not to)?

If I preinstall a `.pth` file in the venv, pointing to the base install, would uv respect that?

**Comment by paveldikov** [comment]
> If I preinstall a .pth file in the venv, pointing to the base install (or some shared libs
folder), would uv respect that?

Just tried that -- it does not, but IMO it should.

**Comment by chrisrodrigue** [comment]
I have a similar use case to @J3ronimo

My org is tied to using vetted Python package distributions such as Anaconda/WinPython in environments that lack access to PyPI.

The general idea is to simultaneously:
- Utilize a base `conda` environment for access to the interpreter (i.e. `python==3.1.4` and the `site-packages` that come bundled with the Anaconda distribution. This is effectively a system python installation with a large number of vetted packages added to it.
- Utilize `uv` with a `pyproject.toml` file as one normally would

IMO, it should be able to detect if `build-system.requires`, `project.dependencies` and `project.optional-dependencies` versions are already satisfied in the environment of the active python interpreter as the default behavior. That way running `uv sync` will not result in inefficiently downloading and installing the dependencies for an online system (or fail outright on an offline system).

**Comment by chrisrodrigue** [comment]
A cop-out way to get this would be to just shell out to `python -m pip list` to see what's already satisfied, but I'm not sure how one would annotate sources in `uv.lock` if they came from elsewhere. Maybe the file system path? (i.e. `C:\ProgramData\anaconda3\Lib\site-packages\<package>`

**Comment by BubuOT** [comment]
https://github.com/astral-sh/uv/issues/7358#issuecomment-2348896632 recommended using `uv pip install --system` as a workaround, but that is not an option for us, because the system python environment is on a read-only partition.

**Comment by awoimbee** [comment]
Same as #6880

**Comment by pradyunsg** [comment]
Would a PR for this be welcome?

> If I preinstall a `.pth` file in the venv, pointing to the base install (or some shared libs folder), would uv respect that?

Currently, no.

> (I think this is roughly a known problem, since we don't look at the full sys.path.)

This is accurate. #3500 pulled in sys.path from the interpreter but it's not being used at this time.

Looking through the code for this, it looks like the changes needed here are:

- Update the environment lookup logic to use sys.path rather than sysconfig paths, for installed distributions.
- Track whether something is coming from "outside" the environment, and skip uninstalling such packages (something that'd fail in many cases, in case of missing permissions).

Based on taking a quick look at things, these aren't particularly complex and I'd be happy to explore doing both these pieces, if a PR for this would be useful/welcome.

**Comment by charliermarsh** [comment]
Definitely welcome. I’m happy to answer questions too.

**Comment by jscanvic** [comment]
@BubuOT Did you end up finding a workaround for read-only system site-packages?

**Comment by ghost** [comment]
@jscanvic not really, no, besides just accepting that uv will re-download and reinstall all the packages that are already in the system venv.

**Comment by anoduck** [comment]
This is a significant issue for users of OpenBSD, exponentially compounded with the current trend to compile everything with rust. This is due to OpenBSD's system security policies not granting user level access to system level libraries. If there is a workaround specific to the platform, I am unaware of it, and have been for some time.

**Comment by frazane** [comment]
This is also an issue for users on HPC systems trying to create python virtual environments that add a few extra dependencies on top of heavy dependencies (e.g. torch) provided by the HPC environment. Currently, I have to switch to `pip` which provides the desired behavior.

**Comment by MeraX** [comment]
I agree with @frazane and their 14 supporters. It would be great to have an option for `uv` to behave like `pip` and respect `--system-site-packages`. This option must not necessarily become the default, but should work upon the users request.
